### PR TITLE
Add mapping Pandas's nullable numeric types

### LINKED ISSUE #2509: Support Python Pandas NA value
Add support for the experimental Python Pandas NA missing value.

https://pandas.pydata.org/docs/user_guide/missing_data.html

### LINKED ISSUE #2512: Mapping pandas wrapped array types (such as FloatArray) led to PyObject being found.
We should check and retrieve the real type (numpy.dtype) and use it to map to DType.

### PR DESCRIPTION
**Pull request by jmao-denver**
This change would recognize the Pandas ExtenstionDtype types and handle the conversion in to_table() according
Fixes #2512
Fixes #2509

### CODE REVIEW

**Review by chipkent — CHANGES_REQUESTED** [review]

**Review by jjbrosnan — COMMENTED** [review]
Second Chip's review

File: py/server/tests/test_pandas.py [hunk 1]
```diff
@@ -165,6 +165,23 @@ def test_to_table_category(self):
...
+        df = pd.DataFrame({"NullableInt64": int_array, "NullableDouble": double_array})
+
+        with self.assertRaises(DHError) as cm:
+            table = to_table(df)
+        self.assertRegex(str(cm.exception), r"TypeError.* NAType")
+
+        int_array = int_array.fillna(NULL_LONG)
+        double_array = double_array.fillna(NULL_DOUBLE)
+        df = pd.DataFrame({"NullableInt64": int_array, "NullableDouble": double_array})
+        table = to_table(df)
```
Comment (chipkent) [inline py/server/tests/test_pandas.py:180]: It seems like `to_table` should be transparently handling the NA conversions without the user having to mess with it.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/server/tests/test_pandas.py (hunk 1 above)
Comment (jmao-denver) [reply py/server/tests/test_pandas.py:180]: I don't oppose to the idea and we have a ticket #2509 for that. To solve would most likely require providing a way for the users to specify replacement values which will be quite a bit more work than what the PR intends to solve.

Comment (chipkent) [reply py/server/tests/test_pandas.py:180]: I'm ok if it is another PR.

**Review by chipkent — DISMISSED** [review]
Approved with the last work to be done in another PR.

File: py/server/tests/test_pandas.py (hunk 1 above)
Comment (jmao-denver) [reply py/server/tests/test_pandas.py]: After discussing it with @rbasralian I went ahead and implemented the transparent handling of pd.NA values. No interface change and it should not preclude us from adding user-supplied custom mapping of pd.NA in the future.

File: py/server/deephaven/constants.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 _JQueryConstants = jpy.get_type("io.deephaven.util.QueryConstants")
 
 # Deephaven Special Null values for primitive types
-NULL_BOOLEAN = _JQueryConstants.NULL_BOOLEAN
+NULL_BOOLEAN = _JQueryConstants.NULL_BYTE
```
Comment (chipkent) [inline py/server/deephaven/constants.py:12]: Is this change intentional?  If so, it seems to deserve a comment.  This makes me nervous without justification.

File: py/server/deephaven/dtypes.py [hunk 3]
```diff
@@ -180,8 +181,18 @@ def from_jtype(j_class: Any) -> DType:
         return dtype
 
 
-def from_np_dtype(np_dtype: np.dtype) -> DType:
-    """ Look up a DType that matches the numpy.dtype, if not found, return PyObject. """
+def from_np_dtype(np_dtype) -> DType:
```
Comment (chipkent) [inline py/server/deephaven/dtypes.py:184]: annotation?

File: py/server/deephaven/numpy.py [hunk 4]
```diff
@@ -164,11 +164,12 @@ def to_table(np_array: np.ndarray, cols: List[str]) -> Table:
                             f"the number of column names {len(cols)}")
 
         input_cols = []
+        dtype = dtypes.from_np_dtype(np_array.dtype)
         if len(cols) == 1:
```
Comment (chipkent) [inline py/server/deephaven/numpy.py:168]:
```suggestion
        dtype = dtypes.from_np_dtype(np_array.dtype)
        
        if len(cols) == 1:
```
Easier to read

File: py/server/deephaven/pandas.py [hunk 5]
```diff
@@ -88,6 +89,33 @@ def to_pandas(table: Table, cols: List[str] = None) -> pandas.DataFrame:
         raise DHError(e, "failed to create a Pandas DataFrame from table.") from e
 
 
+EX_DTYPE_NULL_MAP = {
+    pd.BooleanDtype: NULL_BYTE,
```
Comment (chipkent) [inline py/server/deephaven/pandas.py:93]: NULL_BOOLEAN?

File: py/server/deephaven/pandas.py [hunk 6]
```diff
@@ -112,7 +140,10 @@ def to_table(df: pandas.DataFrame, cols: List[str] = None) -> Table:
 
         input_cols = []
         for col in cols:
-            input_cols.append(_make_input_column(col, df.get(col).values))
+            np_array = df.get(col).values
+            dtype = dtypes.from_np_dtype(np_array.dtype)
+            np_array = _map_na(df.get(col).values)
```
Comment (chipkent) [inline py/server/deephaven/pandas.py:145]:
```suggestion
            np_array = _map_na(np_array)
```

File: py/server/tests/test_pandas.py [hunk 7]
```diff
@@ -165,6 +165,41 @@ def test_to_table_category(self):
...
+        self.assertIs(table.columns[2].data_type, dtypes.int16)
+        self.assertIs(table.columns[3].data_type, dtypes.int32)
+        self.assertIs(table.columns[4].data_type, dtypes.int64)
+        self.assertIs(table.columns[5].data_type, dtypes.float32)
+        self.assertIs(table.columns[6].data_type, dtypes.double)
+        self.assertIs(table.columns[7].data_type, dtypes.string)
+        self.assertEqual(table.size, 3)
+        table_string = table.to_string()
+        self.assertEqual(8, table_string.count("null"))
+
```
Comment (chipkent) [inline py/server/tests/test_pandas.py:202]: What happens if a user has a table like we saw, where a column is `object` and contains strings plus NaNs for missing values?  Are we going to force users to convert to `string` columns?  Yes is probably ok.  I'm just trying to understand the thinking.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/server/deephaven/constants.py (hunk 2 above)
Comment (jmao-denver) [reply py/server/deephaven/constants.py:12]: yes, left a comment

File: py/server/tests/test_pandas.py (hunk 7 above)
Comment (jmao-denver) [reply py/server/tests/test_pandas.py:202]: Yes, Attempting to eagerly infer data type probably has more drawbacks than benefit. Also it is reasonable to ask the users to provide specific data types if they know what they are and don't want to see PyObject columns.

**Review by chipkent — CHANGES_REQUESTED** [review]

**Review by chipkent — DISMISSED** [review]

**Review by jjbrosnan — APPROVED** [review]
